src/utils.rs

```diff
@@ -146,54 +148,84 @@ impl Console {
+                EpollEvent::new(EventSet::IN, self.fd as u64),
+            )
+            .map_err(|e| {
+                new_nitro_cli_failure!(
+                    &format!("Failed to add fd to epoll: {:?}", e),
+                    NitroCliErrorEnum::EpollError
+                )
+            })?;
+
+        // If the function call provides a disconnect timeout, create a timerfd,
+        // arm it and then add it to epoll
+        if let Some(disconnect_timeout) = disconnect_timeout_sec {
+            // Create timerfd
+            let mut timerfd = TimerFd::new().map_err(|e| {
+                new_nitro_cli_failure!(
+                    &format!("Failed to initialize timerfd: {:?}", e),
+                    NitroCliErrorEnum::EpollError
+                )
+            })?;
+
+            // Arm timerfd with disconnect_timeout seconds
+            timerfd
+                .reset(Duration::from_secs(disconnect_timeout), None)
+                .map_err(|e| {
+                    new_nitro_cli_failure!(
+                        &format!("Failed to arm timerfd: {:?}", e),
+                        NitroCliErrorEnum::EpollError
+                    )
+                })?;
+
+            // Add timerfd fd to epoll
+            let timerfd_fd = timerfd.into_raw_fd();
             epoll
-                .ctl(ControlOperation::Add, self.fd, epoll_event)
+                .ctl(
+                    ControlOperation::Add,
+                    timerfd_fd,
+                    EpollEvent::new(EventSet::IN, timerfd_fd as u64),
+                )
                 .map_err(|e| {
                     new_nitro_cli_failure!(
                         &format!("Failed to add fd to epoll: {:?}", e),
                         NitroCliErrorEnum::EpollError
                     )
                 })?;
-
-            // Disconnect timeout in microseconds
-            epoll_timeout_us = (disconnect_timeout_sec.unwrap_or(0) * 1000 * 1000) as i128;
         }
 
         loop {
-            if disconnect_timeout_sec.is_some() {
-                start_epoll_time = Instant::now();
-
-                // Use epoll_wait to exit the blocking state when the fd is ready to be read
-                // or when the disconnect time has passed
-                let num_events = epoll
-                    .wait((epoll_timeout_us / 1000) as i32, &mut events)
-                    .map_err(|e| {
-                        new_nitro_cli_failure!(
-                            &format!("Failed to wait on epoll: {:?}", e),
-                            NitroCliErrorEnum::EpollError
-                        )
-                    })?;
-
-                // If the timeout expires, no event happend and the console disconnects
-                if num_events == 0 {
+            // Wait for kernel notification that one of the fds is available
+            epoll.wait(-1, &mut events).map_err(|e| {
+                new_nitro_cli_failure!(
+                    &format!("Failed to wait epoll: {:?}", e),
```

Comment:
This doesn't seem to properly deal with EINTR/EAGAIN errors. On EINTR/EAGAIN, the operation should be retried. We already have a timeout, so if the timeout is still active, we should not exit here.

Reply:
Thanks for the observation! I have changed the code in a manner that if an interrupt unblocks the `wait` call and the triggered event number is `0`, the loop will just simply continue and block at the next `wait` call.